massa-async-pool/src/pool.rs

```diff
@@ -0,0 +1,178 @@
+    ) -> AsyncPool {
+        AsyncPool {
+            config,
+            messages: snapshot
+                .messages
+                .into_iter()
+                .map(|msg| (msg.compute_id(), msg))
+                .collect(),
+        }
+    }
+
+    /// Returns a snapshot clone of the AsyncPool for bootstrapping other nodes
+    pub fn get_bootstrap_snapshot(&self) -> AsyncPoolBootstrap {
+        AsyncPoolBootstrap {
+            messages: self.messages.values().cloned().collect(),
+        }
+    }
+
+    /// Applies precompiled AsyncPoolChanges to the pool without checking for overflows.
+    /// This function is used when applying pre-compiled AsyncPoolChanges to an AsyncPool.
+    ///
+    /// # arguments
+    /// * changes: AsyncPoolChanges listing all async pool changes (message insertions/deletions)
+    pub fn apply_changes_unchecked(&mut self, changes: AsyncPoolChanges) {
+        for change in changes.0.into_iter() {
+            match change {
+                // add a new message to the pool
+                (msg_id, AddOrDelete::Add(msg)) => {
+                    self.messages.insert(msg_id, msg);
+                }
+
+                // delete a message from the pool
+                (msg_id, AddOrDelete::Delete) => {
+                    self.messages.remove(&msg_id);
+                }
+            }
+        }
+    }
+
+    /// Settles a slot, adding new messages to the pool and returning expired and excess ones.
+    /// This method is called at the end of a slot execution to apply the list of emitted messages,
+    /// and get the list of pruned messages for `coins` reimbursement.
+    ///
+    /// # arguments
+    /// * slot: used to filter out expired messages, not stored
+    /// * new_messages: list of AsyncMessage to add to the pool
+    ///
+    /// # returns
+    /// The list of (message_id, message) that were eliminated from the pool after the changes were applied, sorted in the following order:
+    /// * expired messages from the pool, in priority order (from lowest to highest priority)
+    /// * expired messages from new_messages (in the order they appear in new_messages)
+    /// * excess messages after inserting all remaining new_messages, in priority order (from lowest to highest priority)
+    pub fn settle_slot(
+        &mut self,
+        slot: Slot,
+        new_messages: Vec<AsyncMessage>,
+    ) -> Vec<(AsyncMessageId, AsyncMessage)> {
+        // Compute IDs
+        let mut new_messages: Vec<_> = new_messages
+            .into_iter()
+            .map(|v| (v.compute_id(), v))
+            .collect();
+
+        // Filter out all messages for which the validity end is expired.
+        // Note that the validity_end bound is NOT included in the validity interval of the message.
+        let mut eliminated: Vec<_> = self
+            .messages
+            .drain_filter(|_k, v| slot >= v.validity_end)
+            .chain(new_messages.drain_filter(|(_k, v)| slot > v.validity_end))
```

Comment:
Why this check is different for new messages that for the ones in the async pool ?